packages/fast-components/src/design-system-consumer/design-system-consumer.ts

```diff
@@ -0,0 +1,76 @@
+import { observable } from "@microsoft/fast-element";
+import { DesignSystemResolverEntry } from "../styles/recipes";
+import { composedParent } from "../utilities";
+import { DesignSystemProvider } from "../design-system-provider";
+
+export interface ConsumerArgs {
+    recipes: DesignSystemResolverEntry[];
+}
+
+export interface DesignSystemConsumer {
+    recipes: DesignSystemResolverEntry[];
+    provider: DesignSystemProvider | null;
+}
+
+export function designSystemConsumer<T extends { new (...args: any[]) }>(
+    constructor: T
+): T;
+export function designSystemConsumer<T extends { new (...args: any[]) }>(
+    args: ConsumerArgs
+): <_T>(constructor: _T) => _T;
+export function designSystemConsumer<T extends { new (...args: any[]) }>(
+    argsOrConstructor: any
+): any {
+    function decorator(constructor: T, options: ConsumerArgs): T {
+        class Consumer extends constructor implements DesignSystemConsumer {
+            public readonly recipes = options.recipes;
+
+            @observable
+            public provider: DesignSystemProvider | null = null;
+
+            /**
+             * Find the parent DesignSystem provider.
+             */
+            public findProvider(): DesignSystemProvider | null {
+                let parent = composedParent(this as any);
+
+                while (parent !== null) {
+                    if ((parent as any).isDesignSystemProvider) {
+                        return parent as any;
+                    } else {
+                        parent = composedParent(parent);
+                    }
+                }
+
+                return null;
+            }
+
+            public connectedCallback(): void {
+                super.connectedCallback();
+
+                const provider = this.findProvider();
+
+                if (!!provider) {
+                    this.provider = provider;
+                    this.provider.subscribe(this);
+                }
+            }
+
+            public disconnectedCallback(): void {
+                if (!!this.provider) {
+                    this.provider.unsubscribe(this);
+                }
+            }
+        }
+
+        return Consumer;
+    }
+
+    if (typeof argsOrConstructor === "function") {
+        return decorator(argsOrConstructor, { recipes: [] });
+    } else {
+        return (constructor: T): T => {
+            return decorator(constructor, argsOrConstructor);
+        };
+    }
+}
```

Comment:
Looking at the consumer work, I think we can eventually factor this to work through DI, so we don't need the extra decorator or inheritance. I don't want to block on that because DI just isn't close to ready. So, we should keep this implementation. But, I think we could have something like:

```TypeScript
@customElement({
  name:'foo-bar',
  template,
  styles,
  dependencies: [designSystemConsumer({
    recipes: [
        neutralOutlineRest,
        neutralFillInputRest,
        neutralForegroundRest,
        neutralFillInputHover,
        neutralOutlineHover,
      ]
  })]
})
```

Then FastElement could pull out all Behaviors from the container, one of which would be the consumer. That enables it to instantiate and connect/disconnect along with the element. Basically, this gives Fast Element a sort of mixin system, which can be tapped into through the dependencies configuration.

That's one way we can go. Another way is that if we associate these recipes instead with the styles themsleves, like I think was suggested above. Then since the plumbing is already in place to add styles, and we planned to enable remove with the registry feature...we could also add this logic into that.

Reply:
I think the second option here is ideal. Is that something we have the support for today or do we need to do work in the stylesheet registry to get that to work?